pkg/sql/sem/builtins/pgcrypto_builtins.go

```diff
@@ -161,6 +184,148 @@ var pgcryptoBuiltins = map[string]builtinDefinition{
+	} else if strings.HasPrefix(salt, "$2a$") || strings.HasPrefix(salt, "$2x$") {
+		cryptFunc = cryptBlowfish
+	} else {
+		// todo des/xdes
+		return "", pgerror.New(pgcode.InvalidParameterValue, "invalid salt algorithm")
+	}
+
+	output, err := cryptFunc([]byte(password), []byte(salt))
+	if err != nil {
+		return "", err
+	}
+
+	return string(output), nil
+}
+
+var md5CryptSwaps = [16]int{12, 6, 0, 13, 7, 1, 14, 8, 2, 15, 9, 3, 5, 10, 4, 11}
+
+func cryptMd5(password, salt []byte) ([]byte, error) {
+
+	if len(salt) > 11 {
+		salt = salt[:11] // extra characters are allowed but ignored
+	}
+
+	header := salt[:3]
+	// can be less than 8 characters (including none) even though gen_salt outputs 8
+	randomSalt := salt[3:]
+
+	md5_1 := md5.New()
+
+	md5_1.Write(password)
+	md5_1.Write(header)
+	md5_1.Write(randomSalt)
+
+	md5_2 := md5.New()
+	md5_2.Write(password)
+	md5_2.Write(randomSalt)
+	md5_2.Write(password)
+
+	for i, mixin := 0, md5_2.Sum(nil); i < len(password); i++ {
+		md5_1.Write([]byte{mixin[i%16]})
+	}
+
+	for i := len(password); i != 0; i >>= 1 {
+		if i%2 == 0 {
+			md5_1.Write([]byte{password[0]})
+		} else {
+			md5_1.Write([]byte{0})
+		}
+	}
+
+	final := md5_1.Sum(nil)
+
+	for i := 0; i < md5Rounds; i++ {
+		md5_3 := md5.New()
+		if i&1 == 0 {
+			md5_3.Write(final)
+		} else {
+			md5_3.Write(password)
+		}
+		if i%3 != 0 {
+			md5_3.Write(randomSalt)
+		}
+		if i%7 != 0 {
+			md5_3.Write(password)
+		}
+		if i%2 == 0 {
+			md5_3.Write(password)
+		} else {
+			md5_3.Write(final)
+		}
+		final = md5_3.Sum(nil)
+	}
+
+	output := make([]byte, len(salt)+1+22)
+	copy(output, salt)
+	i := len(salt)
+	output[i] = '$'
+	i++
+	v := uint32(0)
+	bits := uint32(0)
+	for _, swapIndex := range md5CryptSwaps {
+		v |= uint32(final[swapIndex]) << bits
+		for bits = bits + 8; bits > 6; bits -= 6 {
+			output[i] = itoa64[v&0x3f]
+			i++
+			v >>= 6
+		}
+	}
+	output[i] = itoa64[v&0x3f]
+
+	return output, nil
+}
+
+// bcryptLinked bcrypt.bcrypt is private so use go:linkname as a workaround to access it
+//go:linkname bcryptLinked golang.org/x/crypto/bcrypt.bcrypt
+func bcryptLinked(password []byte, cost int, salt []byte) ([]byte, error)
+
+const blowfishMinSaltLength = 29
+
+func cryptBlowfish(password, salt []byte) ([]byte, error) {
+
+	saltLength := len(salt)
+	if saltLength < blowfishMinSaltLength {
+		return nil, errors.WithHintf(
+			pgerror.Newf(pgcode.InvalidParameterValue, "invalid salt length %d", saltLength),
+			"salt must be at least %d characters long",
+			blowfishMinSaltLength,
+		)
+	}
+
+	salt = salt[:29] // extra characters are allowed but ignored
+
+	rounds, err := strconv.Atoi(string(salt[4:6]))
+	if err != nil {
+		return nil, pgerror.Wrap(err, pgcode.InvalidParameterValue, "invalid salt rounds")
+	}
+
+	if err := checkBlowfishRounds(rounds); err != nil {
+		return nil, err
+	}
+
+	if salt[6] != '$' {
+		return nil, pgerror.Newf(
+			pgcode.InvalidParameterValue,
+			"invalid salt format expected $ but got %c",
+			salt[6])
```

Comment:
super nit: i prefer a trailing comma so as to make git blame when more arguments are added not so bad, i.e.

```

   salt[6],
)
```